New article for the Performance API (#69)
xsleaks/wiki: content/docs/attacks/timing-attacks/performance-api.md

title = "Performance API"
description = ""
date = "2020-10-01"
category = [
    "Attack",
]
abuse = [
    "window.performance",
]
defenses = [
    "SameSite Cookies",
    "CORB",
]
menu = "main"
weight = 2
+++
## Performance API
The [`Performance API`](https://developer.mozilla.org/en-US/docs/Web/API/Performance) provides access to performance-related information enhanced by the data from the [`Resource Timing API`](https://developer.mozilla.org/en-US/docs/Web/API/Resource_Timing_API)
which provides the timings of network requests such as the duration but when there’s a `Timing-Allow-Origin: *` header sent by the server the transfer size and domain lookup time is also provided.  
This data can be accessed by using [`performance.getEntries`](https://developer.mozilla.org/en-US/docs/Web/API/Performance/getEntries) or [`performance.getEntriesByName`](https://developer.mozilla.org/en-US/docs/Web/API/Performance/getEntriesByName)
It can also be used to get the execution time using the difference of [`performance.now()`](https://developer.mozilla.org/en-US/docs/Web/API/Performance/now) however this seems to be less precise for a chrome fetch because it only provides the milliseconds.

> ## Get duration of a network request

Reviewer: ```suggestion
## Network duration
```

> {{< hint warning >}} Firefox only provides the milliseconds {{< /hint >}}
> {{< hint important >}} If your using firefox please change the Image() to ```await fetch(href, {mode:"no-cors", credentials: "include"});``` {{< /hint >}}  
> This creates a network request then after 200ms it gets the duration from ```performance.getEntriesByName()```
> ```javascript
> async function getDuration(url) {
>     let href = new URL(url).href;
>     // Using an image instead of fetch() as some requests had duration = 0
>     let image = new Image().src = href;
>     // Wait for request to be added to performance.getEntriesByName();
>     await new Promise(r => setTimeout(r, 200));
>     // Get last added timings
>     let res = performance.getEntriesByName(href).pop();
>     console.log("Request duration: " + res.duration);
>     return res.duration
> }
> ```

Reviewer: ```suggestion
It is possible to retrieve the network duration of a request from the `performance` API.

The below snippet performs a network request then after 200ms it gets the duration from the `performance` object. 

```javascript
async function getNetworkDuration(url) {
    let href = new URL(url).href;
    // Using an image instead of fetch() as some requests had duration = 0
    let image = new Image().src = href;
    // Wait for request to be added to performance.getEntriesByName();
    await new Promise(r => setTimeout(r, 200));
    // Get last added timings
    let res = performance.getEntriesByName(href).pop();
    console.log("Request duration: " + res.duration);
    return res.duration
}

await getNetworkDuration('https://example.org')
```

> ## Detect X-Frame-Options

Reviewer: I feel that the second hint is unclear why it would be needed to replace Image() with fetch, and how exactly could this be done. Could you add more information to it? 

```suggestion
{{< hint info >}} Unlike other browsers, Firefox provides the measurements in milliseconds. {{< /hint >}}

{{< hint tip >}} If you're using Firefox please change the `Image()` to `await fetch(href, {mode:"no-cors", credentials: "include"});` {{< /hint >}}  

## Detecting X-Frame-Options
```

> {{< hint important >}} This technique doesn't seem to work in Firefox {{< /hint >}}

Reviewer: ```suggestion
```

> If a frame embed is blocked it will not be added to performance.getEntries  

Reviewer: ```suggestion
If displaying a page inside an embed (e.g. because of the `X-Frame-Options` header) it will not be added to the `performance` object in Chrome.
```

> If there was a known cached resource on page load firefox could also work.

Reviewer: Could you move the hint after the snippet? I have also hard time understanding this sentence. Could you provide an example for that scenario?

Author: removed sentence as I was for referring to other attacks.

```javascript

> async function ifFrame(url) {

Reviewer: ```suggestion
async function ifFrameBlocked(url) {
```

    let href = new URL(url).href;
    // There may be requests for this url before the function was run.
    let start_count = performance.getEntriesByName(href).length;
    let embed = document.createElement('embed');
    embed.setAttribute("hidden", true);
    embed.src = href;
    document.body.appendChild(embed);
    // Wait for request to be added to performance.getEntriesByName();
    await new Promise(r => setTimeout(r, 200));
    // Remove test embed
    document.body.removeChild(embed)
    return performance.getEntriesByName(href).length > start_count;
}

> ```

Reviewer: ```suggestion

await isFrameBlocked('https://example.org');
```
```

> ## Detect if a redirect is used or cached

Reviewer: ```suggestion

{{< hint note >}} This technique does seem to only work in Chromium based browsers {{< /hint >}}

## Detecting redirects
```

> {{< hint important >}} This technique doesn't seem to work in Firefox {{< /hint >}}

Reviewer: ```suggestion
{{< hint important >}} This technique doesn't seem to work in Firefox {{< /hint >}}
```

Could you move it at the bottom of the section? Does the technique works on other browsers than Chrome or is only Chrome specific?

Author: Moved to bottom of the section.
Have not tested other browsers yet.

Author: Done tests.

> On chrome the duration is negative when theres a redirect.

Reviewer: ```suggestion
It seems that when a redirect occurs, a negative number will be present in Chrome. 
```

```javascript

> async function ifRedirect(url) {  

Reviewer: ```suggestion
async function isRedirected(url) {  
```

    let href = new URL(url).href;

>     await fetch(href, {mode:"no-cors"});

Reviewer: ```suggestion
    await fetch(href, {mode: "no-cors", credentials: "include" });
```

    // Wait for request to be added to performance.getEntriesByName();
    await new Promise(r => setTimeout(r, 200));
    // Get last added timings
    let res = performance.getEntriesByName(href).pop();
    console.log("Request duration: " + res.duration);
    if(res.duration >= 0) return false
    if(res.duration > -10) console.log("Redirect was cached");
    return true;
}

> ```

Reviewer: ```suggestion

await isRedirected('https://google.com'); // should yield a negative number
await isRedirected('https://www.google.com'); // should yield 0
```

```

Author: The function returns true or false.

> # Detect if a resource is cached

Reviewer: ```suggestion
# Detecting cached resources

With the `performance` API it is possible to detect whether a resource was cached or not.
```

> {{< hint important >}} If your using firefox please change the Image() to ```await fetch(href, {mode:"no-cors", credentials: "include"});``` {{< /hint >}}

Reviewer: ```suggestion
{{< hint tip >}} If you're using Firefox please change the `Image()` to `await fetch(href, {mode:"no-cors", credentials: "include"});` {{< /hint >}}
```

Same as before, it isn't clear why the reader should that, and how exactly the snippet should be modified. Also, could you move it at the bottom of the section?

Author: Removed hint as it did not provide enough benefit.

Unless [Cross-Origin Read Blocking]({{< ref "../../defenses/secure-defaults/corb.md" >}}) is triggered (resource is html) the resource will get cached in the processs of the check.  
```javascript
async function ifCached2(url) {
    let href = new URL(url).href;
    // Using an image instead of fetch() as some requests had duration = 0
    let image = new Image().src = href;
    // Wait for request to be added to performance.getEntriesByName();
    await new Promise(r => setTimeout(r, 200));
    // Get last added timings
    let res = performance.getEntriesByName(href).pop();
    console.log("Request duration: " + res.duration);
    // Check if is 304
    if (res.encodedBodySize > 0 && res.transferSize > 0 && res.transferSize < res.encodedBodySize) return true
    if (res.transferSize > 0) return false;
    if (res.decodedBodySize > 0) return true;
    // Use duration if theirs no Timing-Allow-Origin header
    return res.duration < 10;
}
```

> ## Get speed of connection in octets

Reviewer: ```suggestion
## Connection speed

It is possible to measure the speed of the connection in octets.
```

What do you mean by `in octets `?

```javascript
async function getSpeed(count = 10) {
    var total = 0;
    for (let i = 0; i < count; i++) {
        await fetch(location.href, {cache: "no-store"});
        await new Promise(r => setTimeout(r, 200));
        let page = window.performance.getEntriesByName(location.href).pop();
        total += (page.responseEnd - page.responseStart) / page.transferSize;
    }
    return total/count

> }

Reviewer: Could you add comments to the snippet? And an example usage after the function?
